cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AppSystemMetadataWriter.java

```diff
@@ -0,0 +1,114 @@
+ */
+public class AppSystemMetadataWriter extends AbstractSystemMetadataWriter {
+
+  private final ApplicationSpecification appSpec;
+
+  public AppSystemMetadataWriter(MetadataStore metadataStore, Id.Application entityId,
+                                 ApplicationSpecification appSpec) {
+    super(metadataStore, entityId);
+    this.appSpec = appSpec;
+  }
+
+  @Override
+  Map<String, String> getSystemPropertiesToAdd() {
+    ImmutableMap.Builder<String, String> properties = ImmutableMap.builder();
+    addPrograms(properties);
+    addSchedules(properties);
+    // appSpec.getPlugins() returns all instances of all plugins, so there may be duplicates.
+    // we only store unique plugins right now
+    Set<PluginClass> existing = new HashSet<>();
+    for (Plugin plugin : appSpec.getPlugins().values()) {
+      if (!existing.contains(plugin.getPluginClass())) {
+        addPlugin(plugin.getPluginClass(), null, properties);
+        existing.add(plugin.getPluginClass());
+      }
+    }
+    return properties.build();
+  }
+
+  @Override
+  String[] getSystemTagsToAdd() {
+    return new String[] {
+      appSpec.getName(),
+      appSpec.getArtifactId().getName()
+    };
+  }
+
+  private void addPrograms(ImmutableMap.Builder<String, String> properties) {
+    addPrograms(ProgramType.FLOW, appSpec.getFlows().values(), properties);
+    addPrograms(ProgramType.MAPREDUCE, appSpec.getMapReduce().values(), properties);
+    addPrograms(ProgramType.SERVICE, appSpec.getServices().values(), properties);
+    addPrograms(ProgramType.SPARK, appSpec.getSpark().values(), properties);
+    addPrograms(ProgramType.WORKER, appSpec.getWorkers().values(), properties);
+    addPrograms(ProgramType.WORKFLOW, appSpec.getWorkflows().values(), properties);
+  }
+
+  private void addPrograms(ProgramType programType, Iterable<? extends ProgramSpecification> specs,
+                           ImmutableMap.Builder<String, String> properties) {
+    for (ProgramSpecification spec : specs) {
+      properties.put(programType.getPrettyName() + MetadataDataset.KEYVALUE_SEPARATOR + spec.getName(),
+                     spec.getName());
+    }
+  }
+
+  private void addSchedules(ImmutableMap.Builder<String, String> properties) {
+    for (ScheduleSpecification scheduleSpec : appSpec.getSchedules().values()) {
+      Schedule schedule = scheduleSpec.getSchedule();
+      String scheduleExpr;
+      if (schedule instanceof TimeSchedule) {
+        scheduleExpr = ((TimeSchedule) schedule).getCronEntry();
+      } else if (schedule instanceof StreamSizeSchedule) {
+        scheduleExpr = String.valueOf(((StreamSizeSchedule) schedule).getDataTriggerMB());
+      } else {
+        throw new IllegalStateException("Unknown schedule type: " + schedule.getClass().getName());
+      }
+      properties.put("schedule" + MetadataDataset.KEYVALUE_SEPARATOR + schedule.getName(),
+                     scheduleExpr);
```

Comment:
scheduleExp can have a lot of special character http://www.quartz-scheduler.org/documentation/quartz-1.x/tutorials/crontrigger
I don't think its a good idea to store or support query on it. Lets store the schedule description as value if one is provided.